# 01.Importing libraries

In [57]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 02.Importing Data

In [58]:
#shortcut for data files
path = r'C:\Users\Sreelaya Narla\Desktop\PYTHON\2025 Instacart Basket Analysis'
path

'C:\\Users\\Sreelaya Narla\\Desktop\\PYTHON\\2025 Instacart Basket Analysis'

In [59]:
#importing products.csv data set from original data folder
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [60]:
#importing orders_wrangled.csv from prepared data 
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03.Exercise Data consistency checks

## mixed-type data

In [61]:
#Create a Data frame
df_test = pd.DataFrame()

In [62]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [63]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [64]:
#check for mixed type data
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\Sreelaya Narla\AppData\Local\Temp\ipykernel_18332\1282159123.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [65]:
#changing data type
df_test['mix'] = df_test['mix'].astype('str')

## missing values

In [66]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [67]:
#Creating a subset for missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [68]:
#Running explanatory analysis on the df_products dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [69]:
#Fidning the median of the df_prods dataframe
df_prods['prices'].median() 

7.1

In [70]:
#Finding outliers in the df_prods dataframe.
df_prods[df_prods['prices'] > df_prods['prices'].mean() + (2 * df_prods['prices'].std())]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [71]:
#Replace the outliers values in the prices column with the meadian value of 7.1.
df_prods = df_prods.replace({"prices":{14900.0:7.1, 99999.0:7.1}})

In [72]:
#Checking the df_prods dataframe after imputation
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      7.682245
std    14343.717401     38.316774       5.850282      4.200159
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

##### Observation: Explanatory Analysis (Products Dataframe)
After running explanatory analysis on the df_products dataframe, it was discovered that there are outliers present in the data.
The prices column has a mean of 9.99 and a standard deviation of 453.6. This SD is more than 2 standard deviations away from the mean, indicating the presence of outlier values in the data.
FIdentifying all data in the prices column that is 2 standard deviations away from the mean resulted in two entries or rows: prices for 14900 and prices for 99999.
These were identified as outliers and were subsequently replaced with the median value of 7.1.
Given that the minimum value of 1 and maximum value of 99999 produce a wide range of values, it is recommended to replace outliers with the median value.
The median is less sensitive to outliers compared to the mean, making it a more appropriate measure for these extreme values.
After replacing the outliers with the median value, the mean of the dataframe decreased to 7.6.

In [73]:
df_prods.shape

(49693, 5)

In [74]:
#creating new data frame without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [75]:
#verifying the difference in rows
df_prods_clean.shape

(49677, 5)

## duplicates

In [76]:
#Finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [77]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [78]:
##Addressing duplicates
#using shape to verify after we remove duplicates
df_prods_clean.shape

(49677, 5)

In [79]:
#creating a new Data Frame that will not include duplicates by using drop
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [80]:
#verifying the difference in rows
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting the data

In [81]:
#export the dataframe without the index
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index=False)

# TASK 4.5 

## Data consistency checks on orders data frame
### data frame imported above

In [82]:
#Question-2
#Running exploratory analysis for orders dataframe
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [83]:
#Question-3 MIXID DATA TYPES
#Checking for mixed type data
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [84]:
# Running the fucntion to search for a mixed data type in a column in the df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)      

C:\Users\Sreelaya Narla\AppData\Local\Temp\ipykernel_18332\1864822470.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Sreelaya Narla\AppData\Local\Temp\ipykernel_18332\1864822470.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Sreelaya Narla\AppData\Local\Temp\ipykernel_18332\1864822470.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Sreelaya Narla\AppData\Local\Temp\ipykernel_18332\1864822470.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(ax

In [85]:
#using another function i learnt from chatgpt as the one above suggests 
for col in df_ords.columns.tolist():
    weird = df_ords[col].map(type) != type(df_ords[col].iloc[0])
    if weird.any():
        print(col)

order_id
user_id
order_number
orders_day_of_week
order_hour_of_day
days_since_prior_order


In [86]:
# Check unique types in each problematic column
for col in ['order_id', 'user_id', 'order_number', 'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order']:
    print(f"{col} types:", df_ords[col].apply(type).value_counts())


order_id types: order_id
<class 'int'>    3421083
Name: count, dtype: int64
user_id types: user_id
<class 'int'>    3421083
Name: count, dtype: int64
order_number types: order_number
<class 'int'>    3421083
Name: count, dtype: int64
orders_day_of_week types: orders_day_of_week
<class 'int'>    3421083
Name: count, dtype: int64
order_hour_of_day types: order_hour_of_day
<class 'int'>    3421083
Name: count, dtype: int64
days_since_prior_order types: days_since_prior_order
<class 'float'>    3421083
Name: count, dtype: int64


#### Since my current output shows that all types are consistent, I can safely ignore the earlier warning.  — no mixed data types to clean up here!

In [87]:
#Question 5 MISSING VALUES
#Run a check for missing values in the df_ords dataframe.
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##### Why is days_since_prior_order missing?This is expected: The first order a user makes won’t have a "days since previous order" value, because there is no previous order to compare to. So NaN here simply means: "this was the customer's first order."

In [88]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [89]:
#Find out the the missing values based on user_id count summarized 
df_ords[df_ords['days_since_prior_order'].isnull()].groupby('user_id').count().describe()

order_id  order_number  orders_day_of_week  order_hour_of_day  \
count  206209.0      206209.0            206209.0           206209.0   
mean        1.0           1.0                 1.0                1.0   
std         0.0           0.0                 0.0                0.0   
min         1.0           1.0                 1.0                1.0   
25%         1.0           1.0                 1.0                1.0   
50%         1.0           1.0                 1.0                1.0   
75%         1.0           1.0                 1.0                1.0   
max         1.0           1.0                 1.0                1.0   

       days_since_prior_order  
count                206209.0  
mean                      0.0  
std                       0.0  
min                       0.0  
25%                       0.0  
50%                       0.0  
75%                       0.0  
max                       0.0

##### This step summarizes the missing values in the days_since_prior_order column by user ID. The output confirms that each user has exactly one missing entry, which aligns with the assumption that these correspond to each user's first order. It’s a good validation check but not required unless you want to explore missing data patterns more deeply.

#### don’t need to remove or fill in these missing values unless we’re doing a calculation that requires complete data. If so, 
Replace missing values with 0: df_ords['days_since_prior_order'].fillna(0, inplace=True)
(This implies that the first order happened "0 days" after nothing — just a placeholder.)
Or leave them as NaNs if you're filtering for first orders, which can actually be useful information.

In [94]:
#Question- 7 DUPLICATES CHECK
#Finding duplicates
df_dupl = df_ords[df_ords.duplicated()]
df_dupl

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [95]:
#there are no duplicates

# Export the dataframe

In [96]:
#export the dataframe without the index
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index=False)